File: backend/app/services/embedding_index.py
from dataclasses import dataclass

from backend.app.database import (
    CodeChunkEmbeddingRecord,
    CodeChunkRecord,
    CodeChunkSearchHit,
    CodeWikiStore,
)
from backend.app.services.graphrag.utils import batched, embedding_text, stable_id
from backend.app.services.llm.gateway import LLMGateway


@dataclass(frozen=True)
class EmbeddingIndexBuildResult:
    count: int
    model: str


class EmbeddingIndex:
    def __init__(
        self,
        store: CodeWikiStore,
        llm: LLMGateway,
        *,
        batch_size: int = 32,
    ) -> None:
        self.store = store
        self.llm = llm
        self.batch_size = batch_size

    async def build(
        self,
        repo_id: str,
        chunks: list[CodeChunkRecord],
    ) -> EmbeddingIndexBuildResult:
        model = self.model
        unique_chunks = _dedupe_chunks_by_content_hash(chunks)
        vectors_by_hash = {
            embedding.content_hash: embedding.embedding
            for embedding in self.store.list_code_chunk_embeddings(repo_id, model=model)
            if embedding.embedding
        }
        missing_chunks = [
            chunk for chunk in unique_chunks
            if chunk.content_hash not in vectors_by_hash
        ]
        for batch in batched(missing_chunks, self.batch_size):
            texts = [embedding_text(chunk) for chunk in batch]
            vectors = await self.llm.embed(texts, task_type="embedding")
            for chunk, vector in zip(batch, vectors, strict=True):
                vectors_by_hash[chunk.content_hash] = vector

        records: list[CodeChunkEmbeddingRecord] = []
        for chunk in chunks:
            vector = vectors_by_hash[chunk.content_hash]
            records.append(
                CodeChunkEmbeddingRecord(
                    id=stable_id(repo_id, "embedding", model, chunk.id, chunk.content_hash),
                    repo_id=repo_id,
                    chunk_id=chunk.id,
                    model=model,
                    dimensions=len(vector),
                    embedding=vector,
                    content_hash=chunk.content_hash,
                    created_at=None,
                )
            )
        self.store.sync_code_chunk_embeddings(repo_id, model=model, embeddings=records)
        return EmbeddingIndexBuildResult(count=len(records), model=model)

    async def ensure(
        self,
        repo_id: str,
        chunks: list[CodeChunkRecord],
    ) -> EmbeddingIndexBuildResult | None:
        model = self.model
        if not chunks:
            return None
        existing_chunk_ids = {
            embedding.chunk_id
            for embedding in self.store.list_code_chunk_embeddings(repo_id, model=model)
        }
        if {chunk.id for chunk in chunks} <= existing_chunk_ids:
            return None
        return await self.build(repo_id, chunks)

    async def search(
        self,
        repo_id: str,
        query: str,
        chunks: list[CodeChunkRecord],
        *,
        limit: int,
    ) -> list[CodeChunkSearchHit]:
        await self.ensure(repo_id, chunks)
        vectors = await self.llm.embed([query], task_type="embedding")
        if not vectors:
            return []
        return self.store.search_code_chunk_embeddings(
            repo_id,
            model=self.model,
            query_embedding=vectors[0],
            limit=limit,
        )

    @property
    def model(self) -> str:
        return self.llm.router.profile_for("embedding").model


async def embed_chunks(
    store: CodeWikiStore,
    llm: LLMGateway,
    repo_id: str,
    chunks: list[CodeChunkRecord],
) -> tuple[int, str]:
    result = await EmbeddingIndex(store, llm).build(repo_id, chunks)
    return result.count, result.model


def _dedupe_chunks_by_content_hash(chunks: list[CodeChunkRecord]) -> list[CodeChunkRecord]:
    unique_chunks: list[CodeChunkRecord] = []
    seen_hashes: set[str] = set()
    for chunk in chunks:
        if chunk.content_hash in seen_hashes:
            continue
        seen_hashes.add(chunk.content_hash)
        unique_chunks.append(chunk)
    return unique_chunks

File: backend/src/graphrag/embeddingIndex.ts
import { digest } from "../analysis/graphUtils.js";
import type { CodeWikiStoreApi } from "../db/types.js";
import { validationError } from "../errors.js";
import { providerUserIdForRepo } from "../llm/cache.js";
import type { LlmEmbeddingGateway } from "../llm/gateway.js";
import type { CodeChunk, CodeChunkEmbedding } from "../types.js";

export type EmbeddingIndexBuildResult = {
  count: number;
  model: string;
};

export type CodeChunkHit = {
  chunk: CodeChunk;
  score: number;
  match_type: string;
};

export class EmbeddingIndex {
  constructor(
    private readonly store: CodeWikiStoreApi,
    private readonly provider: LlmEmbeddingGateway,
    private readonly batchSize = 32,
  ) {}

  async build(
    repoId: string,
    chunks: CodeChunk[],
  ): Promise<EmbeddingIndexBuildResult> {
    const model = this.embeddingModel();
    const vectorsByHash = await this.vectorsByContentHash(repoId, model);
    const missingChunks = uniqueChunksByContentHash(chunks).filter(
      (chunk) => !vectorsByHash.has(chunk.content_hash),
    );

    for (const batch of batches(missingChunks, this.batchSize)) {
      const result = await this.provider.embed(
        "embedding",
        batch.map(embeddingText),
        { providerUserId: providerUserIdForRepo(repoId) },
      );
      if (result.embeddings.length !== batch.length) {
        throw validationError(
          `Embedding provider returned ${result.embeddings.length} vectors for ${batch.length} chunks.`,
        );
      }
      batch.forEach((chunk, index) => {
        vectorsByHash.set(chunk.content_hash, result.embeddings[index]!);
      });
    }

    const embeddings = chunks.flatMap((chunk) => {
      const vector = vectorsByHash.get(chunk.content_hash);
      return vector ? [embeddingRecord(repoId, model, chunk, vector)] : [];
    });
    this.store.replaceCodeChunkEmbeddings(repoId, { model, embeddings });
    return { count: embeddings.length, model };
  }

  async ensure(
    repoId: string,
    chunks: CodeChunk[],
  ): Promise<EmbeddingIndexBuildResult | null> {
    const model = this.embeddingModel();
    const existingChunkIds = new Set(
      this.store
        .listCodeChunkEmbeddings(repoId, { model })
        .map((embedding) => embedding.chunk_id),
    );
    if (
      chunks.length > 0 &&
      chunks.every((chunk) => existingChunkIds.has(chunk.id))
    ) {
      return null;
    }
    return this.build(repoId, chunks);
  }

  async search(
    repoId: string,
    query: string,
    chunks: CodeChunk[],
    limit: number,
  ): Promise<CodeChunkHit[]> {
    await this.ensure(repoId, chunks);
    const model = this.embeddingModel();
    const result = await this.provider.embed("embedding", [query], {
      providerUserId: providerUserIdForRepo(repoId),
    });
    const [queryEmbedding] = result.embeddings;
    if (!queryEmbedding) {
      return [];
    }
    return this.store.searchCodeChunkEmbeddings(repoId, {
      model,
      queryEmbedding,
      limit,
    });
  }

  private embeddingModel(): string {
    if (!this.provider.isConfigured("embedding")) {
      throw validationError(
        "Embedding search requires CODEWIKI_LLM__PROFILES__EMBEDDING__MODEL plus credentials or an endpoint.",
      );
    }
    return this.provider.profile("embedding").model;
  }

  private async vectorsByContentHash(
    repoId: string,
    model: string,
  ): Promise<Map<string, number[]>> {
    return new Map(
      this.store
        .listCodeChunkEmbeddings(repoId, { model })
        .filter((embedding) => embedding.embedding.length > 0)
        .map((embedding) => [embedding.content_hash, embedding.embedding]),
    );
  }
}

function embeddingRecord(
  repoId: string,
  model: string,
  chunk: CodeChunk,
  vector: number[],
): CodeChunkEmbedding {
  return {
    id: digest(
      [repoId, "embedding", model, chunk.id, chunk.content_hash].join("\0"),
    ),
    repo_id: repoId,
    chunk_id: chunk.id,
    model,
    dimensions: vector.length,
    embedding: vector,
    content_hash: chunk.content_hash,
    created_at: null,
  };
}

function embeddingText(chunk: CodeChunk): string {
  return [
    `File: ${chunk.file_path}`,
    `Lines: ${chunk.start_line}-${chunk.end_line}`,
    chunk.content,
  ].join("\n");
}

function uniqueChunksByContentHash(chunks: CodeChunk[]): CodeChunk[] {
  const seen = new Set<string>();
  const result: CodeChunk[] = [];
  for (const chunk of chunks) {
    if (seen.has(chunk.content_hash)) {
      continue;
    }
    seen.add(chunk.content_hash);
    result.push(chunk);
  }
  return result;
}

function batches<T>(items: T[], size: number): T[][] {
  const normalizedSize = Math.max(1, size);
  const result: T[][] = [];
  for (let index = 0; index < items.length; index += normalizedSize) {
    result.push(items.slice(index, index + normalizedSize));
  }
  return result;
}
